### PR TITLE
Feature/deep solver

### PR DESCRIPTION
**Pull request by DoubleTimeOnly**
## Summary
Implement a neural network based solver for predicting the previous generation given a board state.
The network is trained on randomly generated board states.
Loss is computed as follows
1. Generate a random board state
2. Compute next generation of board state. Now we have a ground truth pair of previous and next generation.
3. Have model predict the original board state given the next generation

The model performs pretty poorly and on average has 30 bits of error on a 10x10 board.

Future ideas for improvement
* Better loss representation: Multiple previous generations can lead to the same current generation. The current loss penalizes guess of the previous generation that isn't the one we passed in, even if the guess would still lead to a close next generation. Maybe we could try some loss that compares the next generation instead.
* Better model architecture: We currently use a very basic MLP.

### CODE REVIEW

**Comment by amarzot** [comment]
Preface: I haven't looked at the code.

> Maybe we could try some loss that compares the next generation instead

I think this is a great idea. GD algo uses mse of the guess after one gen and the target board. A board state should be scored based on how close it is after one generation. I think that makes sense, as that's really what we care about

**Comment by DoubleTimeOnly** [comment]
Main issue is that the conway forward pass requires hard thresholds (i.e. 1 if neighbors == 3). Which is non differntiable. Might have to think of a slight variant on that.

**Comment by DoubleTimeOnly** [comment]
Another idea is training a model to estimate how close a given board state is to a target board state and use that network to choose the next step in gradient descent solver (instead of measuring board error as the selection metric).
